File: polctrl/python/fixedpoint.py
"""
Fixed-point arithmetic: Q8.8 format, mirroring c/src/fixedpoint.c exactly.

All operations produce bit-identical results to the C implementation,
including overflow/wrapping and rounding behavior.

Conventions (must match C):
  - fp_mul: arithmetic right shift (>> on signed int = truncation toward -inf)
  - fp_div: C integer division (truncation toward zero)
  - Overflow: clamp to int16_t range [-32768, 32767] after each operation
"""

import struct

FP_SHIFT = 8
FP_ONE = 1 << FP_SHIFT
FP_MAX = 32767
FP_MIN = -32768


def _to_int16(x):
    """Simulate int16_t: mask to 16 bits and sign-extend."""
    x = x & 0xFFFF
    if x >= 0x8000:
        x -= 0x10000
    return x


def _to_int32(x):
    """Simulate int32_t: mask to 32 bits and sign-extend."""
    x = x & 0xFFFFFFFF
    if x >= 0x80000000:
        x -= 0x100000000
    return x


def _c_div(a, b):
    """
    Integer division truncating toward zero (matching C's / operator).
    Python's // truncates toward -inf, so we need this wrapper.
    """
    if b == 0:
        raise ZeroDivisionError
    q = abs(a) // abs(b)
    if (a < 0) != (b < 0):
        return -q
    return q


def fp_from_float(x: float) -> int:
    """Convert float to Q8.8. ONLY for tests/initialization."""
    scaled = x * FP_ONE
    # Round to nearest (matching C implementation)
    if scaled >= 0:
        scaled = int(scaled + 0.5)
    else:
        scaled = int(scaled - 0.5)
    if scaled > FP_MAX:
        scaled = FP_MAX
    if scaled < FP_MIN:
        scaled = FP_MIN
    return _to_int16(scaled)


def fp_to_float(x: int) -> float:
    """Convert Q8.8 to float. ONLY for tests/debugging."""
    return x / FP_ONE


def fp_mul(a: int, b: int) -> int:
    """
    Multiply two Q8.8 values. Uses int32_t intermediate.
    Arithmetic right shift (truncation toward -inf), matching C >>.
    """
    temp = _to_int32(a * b)
    # Python's >> on negative integers does arithmetic shift (toward -inf),
    # matching C's >> on signed integers in gcc.
    temp = temp >> FP_SHIFT
    if temp > FP_MAX:
        temp = FP_MAX
    if temp < FP_MIN:
        temp = FP_MIN
    return _to_int16(temp)


def fp_div(a: int, b: int) -> int:
    """
    Divide two Q8.8 values. Uses int32_t intermediate.
    C integer division (truncation toward zero).
    Division by zero returns saturating value.
    """
    if b == 0:
        if a > 0:
            return FP_MAX
        if a < 0:
            return FP_MIN
        return 0

    temp = _to_int32((a << FP_SHIFT))
    temp = _c_div(temp, b)

    if temp > FP_MAX:
        temp = FP_MAX
    if temp < FP_MIN:
        temp = FP_MIN
    return _to_int16(temp)


def fp_clamp(x: int, lo: int, hi: int) -> int:
    """Clamp x to [lo, hi]."""
    if x < lo:
        return lo
    if x > hi:
        return hi
    return x


def fp_abs(x: int) -> int:
    """Absolute value. Guards against FP_MIN overflow."""
    if x < 0:
        if x == FP_MIN:
            return FP_MAX
        return -x
    return x

File: c/src/fixedpoint.c
#include "fixedpoint.h"

/*
 * Q8.8 fixed-point arithmetic implementation.
 *
 * All arithmetic uses int32_t as intermediate type to prevent overflow
 * during multiplication and division, then results are clamped back to
 * int16_t (fp_t).
 *
 * Rounding convention:
 *   - fp_mul: arithmetic right shift (truncation toward -inf).
 *     This matches Python's >> operator on signed integers.
 *   - fp_div: C integer division (truncation toward zero).
 *     Python mirror uses explicit truncation-toward-zero to match.
 *
 * Both choices are fully deterministic and produce bit-identical results
 * between C and Python.
 */

fp_t fp_from_float(float x)
{
    float scaled = x * (float)FP_ONE;
    /* Round to nearest */
    if (scaled >= 0.0f) {
        scaled += 0.5f;
    } else {
        scaled -= 0.5f;
    }
    if (scaled > (float)FP_MAX) scaled = (float)FP_MAX;
    if (scaled < (float)FP_MIN) scaled = (float)FP_MIN;
    return (fp_t)scaled;
}

float fp_to_float(fp_t x)
{
    return (float)x / (float)FP_ONE;
}

fp_t fp_mul(fp_t a, fp_t b)
{
    int32_t temp = (int32_t)a * (int32_t)b;
    /*
     * Arithmetic right shift (>> on signed int in gcc = sign-extending).
     * This truncates toward negative infinity, matching Python's >>.
     * Result range: temp is at most ~32767*32767 ~= 1.07e9, fits int32_t.
     * After >> 8: at most ~4.19e6, must clamp to int16_t.
     */
    temp = temp >> FP_SHIFT;
    if (temp > FP_MAX) temp = FP_MAX;
    if (temp < FP_MIN) temp = FP_MIN;
    return (fp_t)temp;
}

fp_t fp_div(fp_t a, fp_t b)
{
    int32_t temp;

    /* Division by zero: return saturating value */
    if (b == 0) {
        if (a > 0)  return FP_MAX;
        if (a < 0)  return FP_MIN;
        return 0;
    }

    /*
     * (a << FP_SHIFT) / b in Q8.8.
     * a is int16_t [-32768, 32767], << 8 gives [-8388608, 8388608], fits int32_t.
     * C division truncates toward zero.
     */
    temp = ((int32_t)a << FP_SHIFT) / (int32_t)b;

    if (temp > FP_MAX) temp = FP_MAX;
    if (temp < FP_MIN) temp = FP_MIN;
    return (fp_t)temp;
}

fp_t fp_clamp(fp_t x, fp_t lo, fp_t hi)
{
    if (x < lo) return lo;
    if (x > hi) return hi;
    return x;
}

fp_t fp_abs(fp_t x)
{
    if (x < 0) {
        /* Guard against FP_MIN (-32768) where -x would overflow */
        if (x == FP_MIN) return FP_MAX;
        return (fp_t)(-x);
    }
    return x;
}
